vault/auth.go

```diff
@@ -499,50 +500,210 @@ func (c *Core) taintCredEntry(ctx context.Context, nsID, path string, updateStor
+	}
+
+	var needPersist bool
+	if len(globalEntries) == 0 {
+		c.logger.Info("no auth mounts in transactional auth mount table; adding default auth mount table")
+		c.auth = c.defaultAuthTable()
+		needPersist = true
+	} else {
+		c.auth = &MountTable{
+			Type: credentialTableType,
+		}
+
+		for index, uuid := range globalEntries {
+			entry, err := c.fetchAndDecodeMountTableEntry(ctx, barrier, coreAuthConfigPath, uuid)
+			if err != nil {
+				return fmt.Errorf("error loading auth mount table entry (%v/%v): %w", index, uuid, err)
+			}
+
+			if entry != nil {
+				c.auth.Entries = append(c.auth.Entries, entry)
+			}
+		}
+	}
+
+	if len(localEntries) > 0 {
+		for index, uuid := range localEntries {
+			entry, err := c.fetchAndDecodeMountTableEntry(ctx, barrier, coreLocalAuthConfigPath, uuid)
+			if err != nil {
+				return fmt.Errorf("error loading local auth mount table entry (%v/%v): %w", index, uuid, err)
+			}
+
+			if entry != nil {
+				c.auth.Entries = append(c.auth.Entries, entry)
+			}
+		}
+	}
+
+	err = c.runCredentialUpdates(ctx, barrier, needPersist)
+	if err != nil {
+		c.logger.Error("failed to run legacy auth mount table upgrades", "error", err)
+		return err
+	}
+
+	return nil
+}
+
+// This function reads the legacy, single-entry combined auth mount table,
+// returning true if it was used. This will let us know (if we're inside
+// a transaction) if we need to do an upgrade.
+func (c *Core) loadLegacyCredentials(ctx context.Context, barrier logical.Storage) (bool, error) {
+	// Load the existing auth mount table
+	raw, err := barrier.Get(ctx, coreAuthConfigPath)
 	if err != nil {
 		c.logger.Error("failed to read auth table", "error", err)
-		return errLoadAuthFailed
+		return false, errLoadAuthFailed
 	}
-	rawLocal, err := c.barrier.Get(ctx, coreLocalAuthConfigPath)
+	rawLocal, err := barrier.Get(ctx, coreLocalAuthConfigPath)
 	if err != nil {
 		c.logger.Error("failed to read local auth table", "error", err)
-		return errLoadAuthFailed
+		return false, errLoadAuthFailed
 	}
 
-	c.authLock.Lock()
-	defer c.authLock.Unlock()
-
 	if raw != nil {
 		authTable, err := c.decodeMountTable(ctx, raw.Value)
 		if err != nil {
 			c.logger.Error("failed to decompress and/or decode the auth table", "error", err)
-			return err
+			return false, err
 		}
 		c.auth = authTable
 	}
 
 	var needPersist bool
 	if c.auth == nil {
+		// In the event we are inside a transaction, we do not yet know if
+		// we have a transactional mount table; exit early and load the new format.
+		if _, ok := barrier.(logical.Transaction); ok {
+			return false, nil
+		}
+		c.logger.Info("no mounts in legacy auth table; adding default mount table")
 		c.auth = c.defaultAuthTable()
 		needPersist = true
-	} else {
-		// only record tableMetrics if we have loaded something from storge
-		c.tableMetrics(len(c.auth.Entries), false, true, raw.Value)
+	} else if _, ok := barrier.(logical.Transaction); ok {
```

Comment:
the legacy path should be similar to what it was before tx aware backend - is the else dropped here necessary?

Reply:
Ah, we should probably make sure we emit metrics, good catch.

Otherwise, the else-if is necessary to ensure we write the legacy entries to the new format.